### PR TITLE
[NF] Add a Deprecation system

### PR DESCRIPTION
**Pull request by skoudoro**
The goal of this PR is to bring a standard way to deprecate functions, modules, ... 

This is heavily based on the Nibabel deprecation system and adds a dependency to the small package named `packaging`. Usually, you do not need to install this package because `setuptools` depends on it and already did the installation for you. But it can happen that `setuptools` is not installed in some environments.

### CODE REVIEW

**Comment by skoudoro** [comment]
Any opinion concerning this PR?

File: dipy/testing/__init__.py [hunk 1]
```diff
@@ -43,8 +43,76 @@ def assert_arrays_equal(arrays1, arrays2):
...
+    Examples
+    --------
+    >>> import warnings
+    >>> with clear_and_catch_warnings(modules=[np.core.fromnumeric]):
+    ...     warnings.simplefilter('always')
+    ...     # do something that raises a warning in np.core.fromnumeric
+
+    Note
+    ----
+    this class is copied (with minor modifications) from the Nibabel.
```
Comment (Garyfallidis) [inline dipy/testing/__init__.py]: This class is copied (with minor modifications) from Nibabel

File: dipy/utils/tests/test_deprecator.py [hunk 2]
```diff
@@ -0,0 +1,169 @@
+"""Module to test ``dipy.utils.deprecator`` module.
+
+Notes
+-----
+this file is copied (with minor modifications) from the Nibabel.
```
Comment (Garyfallidis) [inline dipy/utils/tests/test_deprecator.py]: Same as before

**Comment by Garyfallidis** [comment]
So ``package`` will be installed if not installed in the system?

**Comment by skoudoro** [comment]
> So package will be installed if not installed in the system?

Yes, for all future release.

Concerning the current master, you may have to install it. But I suppose it is already done since you need `setuptools` to install DIPY and `setuptools` depends on `packaging`

**Comment by Garyfallidis** [comment]
@arokem can you check this PR too?

**Comment by arokem** [comment]
Yeah. LGTM!

**Pull request merged**